numba/npyufunc/parfor.py

```diff
@@ -70,6 +72,110 @@ def _lower_parfor_parallel(lowerer, parfor):
+
+            # Create var for outer dimension size of reduction array equal to number of threads.
+            num_threads_var = ir.Var(scope, mk_unique_var("num_threads"), loc)
+            num_threads_assign = ir.Assign(ir.Const(thread_count, loc), num_threads_var, loc)
+            typemap[num_threads_var.name] = types.intp
+            lowerer.lower_inst(num_threads_assign)
+
+            # Empty call takes tuple of sizes.  Create here and fill in outer dimension (num threads).
+            size_var = ir.Var(scope, mk_unique_var("tuple_size_var"), loc)
+            typemap[size_var.name] = types.UniTuple(types.intp, redarrdim)
+            size_var_list = [num_threads_var]
+
+            # If this is a reduction over an array...
+            if isinstance(redvar_typ, types.npytypes.Array):
+                # Add code to get the shape of the array being reduced over.
+                redshape_var = ir.Var(scope, mk_unique_var("redarr_shape"), loc)
+                typemap[redshape_var.name] = types.UniTuple(types.intp, redvar_typ.ndim)
+                redshape_getattr = ir.Expr.getattr(redvar, "shape", loc)
+                redshape_assign = ir.Assign(redshape_getattr, redshape_var, loc)
+                lowerer.lower_inst(redshape_assign)
+
+                # Add the dimension sizes of the array being reduced over to the tuple of sizes pass to empty.
+                for j in range(redvar_typ.ndim):
+                    onedimvar = ir.Var(scope, mk_unique_var("redshapeonedim"), loc)
+                    onedimgetitem = ir.Expr.static_getitem(redshape_var, j, None, loc)
+                    typemap[onedimvar.name] = types.intp
+                    onedimassign = ir.Assign(onedimgetitem, onedimvar, loc)
+                    lowerer.lower_inst(onedimassign)
+                    size_var_list += [onedimvar]
+
+            size_call = ir.Expr.build_tuple(size_var_list, loc)
+            size_assign = ir.Assign(size_call, size_var, loc)
+            lowerer.lower_inst(size_assign)
+
+            #size_print = ir.Print(args=[size_var], vararg=None, loc=loc)
+            #lowerer.fndesc.calltypes[size_print] = signature(types.none, typemap[size_var.name])
+            #lowerer.lower_inst(size_print)
+
+            # Add call to empty passing the size var tuple.
+            full_call = ir.Expr.call(empty_func, [size_var], {}, loc=loc)
+            redarr_var = ir.Var(scope, mk_unique_var("redarr"), loc)
+            typemap[redarr_var.name] = redarrvar_typ
+            full_call_assign = ir.Assign(full_call, redarr_var, loc)
+            lowerer.fndesc.calltypes[full_call] = ff_sig
+            lowerer.lower_inst(full_call_assign)
+
+            # Remember mapping of original reduction array to the newly created per-worker reduction array.
+            redarrs[redvar.name] = redarr_var
+
+            init_val = parfor_reddict[parfor_redvars[i]][0]
+            if init_val != None and not isinstance(redvar_typ, types.npytypes.Array):
```

Comment:
Array reductions should have the same behavior as scalars: for known reductions (+, *) the initial value should be set automatically instead of relying on the user.

Reply:
Done.